wrappers/core18.go

```diff
@@ -657,6 +661,50 @@ func undoSnapdDbusActivationOnCore() error {
 	return err
 }
 
+func snapdDesktopFileNames(s *snap.Info) ([]string, error) {
+	paths, err := filepath.Glob(filepath.Join(s.MountDir(), "usr/share/applications/*.desktop"))
+	if err != nil {
+		return nil, err
+	}
+
+	fileNames := make([]string, len(paths))
+	for i, path := range paths {
+		fileNames[i] = filepath.Base(path)
+	}
+
+	return fileNames, nil
+}
+
+func writeSnapdDesktopFilesOnCore(s *snap.Info) error {
+	// Ensure /var/lib/snapd/desktop/applications exists
+	if err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755); err != nil {
+		return err
+	}
+
+	desktopFileNames, err := snapdDesktopFileNames(s)
+	if err != nil {
+		return err
+	}
+
+	desktopFiles := make(map[string]osutil.FileState, len(desktopFileNames))
+	for _, fileName := range desktopFileNames {
+		desktopFiles[fileName] = &osutil.FileReference{Path: filepath.Join(s.MountDir(), "usr/share/applications", fileName)}
```

Comment:
Small question, Is it intentional that `snapd` related desktop files don't have the prefix that other snaps have (i.e. `<s.DesktopPrefix()>_<desktop-file name>`)?

It is not very useful during addition but during removal it will be useful if any of the desktop files are removed from `snapd`'s source, as without using the `<s.DesktopPrefix()>_*` glob the old file will not be removed.

For context: https://github.com/snapcore/snapd/blob/cd03bb167398bf86a11998511a1f2f07a366f8d6/wrappers/desktop.go#L292

Reply:
@jhenstridge perhaps you know about if these files should/shouldn't have prefixes?

Reply:
We need the `io.snapcraft.SessionAgent.desktop` to have that name so that it matches the D-Bus name claimed by the session agent. This is a hard requirement for use of the newer GTK notifications API, and the prime motivation for this PR as I understand it. A D-Bus name cannot include underscores, so the naming snapd uses for snap desktop files is not usable.

And at least for now, this guarantees that these desktop files can't conflict with other snap desktop files.

Reply:
@jhenstridge Thank you for the explanation!